### DISCUSSION TITLE
zero feed is not working as expected, missing zero export script switches

### DISCUSSION THREAD

**Discussion by rradkov** [hoylabs/OpenDTU-OnBattery#1305] [General]
Hello everyone !
first of all I want to thanks to everyone who contribute to this project to be so great!

Unfortunately I have a few issues with OpenDTU DPL policies which lead me to return to the grid more than 1kW of energy for the past month.
Here is my setup:
I have generic esp32s3_usb with 8MB flash memory. Firmware 2024.09.11
HMS-2000 4T inverter.
1 Shelly EM for monitoring import/export from/to the grid
1 shelly 1pm for monitoring the HMS and in case of DTU failure to turn off the inverter.
In the beginning I was with OpenDTU and desktop device with purpose of running zero export script.
The main problem then was that the OpenDTU was not able to reconnect to the HMS each morning. Temporary solution was each morning to power cycle the DTU.
I've asked for help in OpenDTU community and they recommended to reinstall the ESP with OpenDTU-OnBattery.
In the begging all was good, DTU was able to connect to the inverter without any issues but lately I'm noticing that the configuration and settings regarding zero export is not working as expected. For the past month(September) total energy returned is more than 1,7kW which I think is a lot !
![image](https://github.com/user-attachments/assets/ebfede0a-3c80-4e83-bdd1-a99118cd9907)
I know that usually when applying new settings to the inverter it can take up to 3 seconds but lately I'm noticing that there are more that 5 seconds before new settings is applied to the inverter. Here is few screenshots from Home assistance as evidence
![image](https://github.com/user-attachments/assets/b4708fa2-2f98-4998-92c3-95ca4c9affc7)
![image](https://github.com/user-attachments/assets/0f973b4a-a314-45ce-ad75-70a2de7c7168)

Here is my current configuration:
DTU each second is gathering the information from my Shelly EM
![image](https://github.com/user-attachments/assets/a31f8479-24a0-4b89-8f48-3a6a079244e4)

Here is the settings for the inverter:
![image](https://github.com/user-attachments/assets/0584c9ee-4015-4e15-a18e-a2ce88170c46)

Here is the pulling interval
![image](https://github.com/user-attachments/assets/96952cff-9a6c-4b6c-8d3d-3cf4adc223f4)

Can someone advice me which settings I should configure in order to reduce exported energy ?

Some of the settings that I'm missing here that can potentially "fix" this issue from Zero export script are:
Intermediate meter
POWERMETER_MIN_POINT
ON_GRID_FEED_FAST_LIMIT_DECREASE

Are you planning implementing those two settings soon?

**Comment by Gururom** [comment]
Deine URL Zeile zur Abfrage des Shellys 3EM ist meiner meinung nach falsch.
Da sollte stehen "http://ShellyIP/status"
Aber dafür gibts je den Testbutton unten

**Reply by spcqike** [reply]
kann ich mir nicht erklären, da die 0-Einspeisung tendenziell geht. er regelt ja korrekt aus. URL hin oder her. (es gibt sicher mehrer möglichkeiten :) )

Ich sehe auf der anderen Seite aber auch kein echtes problem. 1.7kWh im Monat? 0.3% des monatlichen Verbrauchs..... ich weiß nicht, ob man sich darüber aufregen kann oder muss.

irgendein Verbraucher scheint um ~600W zu schwingen. bei einer Regelzeit von 5s, also 5s Einspeisung von Überschuss, gehen 0.83Wh in den Export. das macht bei 1.7kWh monatlichem Export halt ~2,000 Regelzyklen aus. das sind 66 Zyklen / Überschwinger am Tag bzw. nicht mal 3 pro Stunde. Wie man aber sieht, sind es weit mehr.

man kann versuchen die Zykluszeiten auf 1s zu stellen, also DTU und powermeter. viel rausholen wird das nicht. meiner Meinung nach muss man das hier aber auch nicht. (wir reden hier von "verschenkten" ~60 Cent im Monat)


was mir jedoch auffällt: das minimale limit (10W) und die Baseload (100W) sind für den HMS-2000 zu gering. das erklärt zwar nicht sein "Exportproblem", ist aber halt auch nicht ideal.

**Reply by rradkov** [reply]
> Your URL line for querying the Shellys 3EM is wrong in my opinion. It should say " http://ShellyIP/status " But there is a test button below for that

URL is correct I've tested it multiple times.
![image](https://github.com/user-attachments/assets/42921f5f-cf2a-47ca-a0d9-926b1f49baff)


> I can't explain it, because the 0 input tends to work. It regulates correctly. URL or not. (there are certainly more possibilities :) )
> 
> On the other hand, I don't see any real problem. 1.7kWh a month? 0.3% of monthly consumption..... I don't know if you can or should get upset about that.
> 
> Some consumer seems to be oscillating around ~600W. With a control time of 5s, i.e. 5s of surplus feed-in, 0.83Wh is exported. With 1.7kWh of monthly export, that amounts to ~2,000 control cycles. That's 66 cycles/overshoots per day, or not even 3 per hour. But as you can see, there are far more.
> 
> You can try setting the cycle times to 1s, i.e. DTU and power meter. That won't get you much out of it. But in my opinion you don't need to do that here. (We're talking about "wasted" ~60 cents a month here)
> 
> What I notice, however, is that the minimum limit (10W) and the base load (100W) are too low for the HMS-2000. That doesn't explain its "export problem", but it's not ideal either.

Unfortunately here in Bulgaria if you export energy to the grid, you need to follow some strict regulatory laws(only certified electrician can install PV systems) and sign a contract with electricity service provider. If PV system is not certified and you're exporting to the grid, I need to pay exported energy at the price of importing it. Also they can potentially give me a ticket for that.

**Reply by schlimmchen** [reply]
Sorry, but OpenDTU-OnBattery is a hobbyist/enthusiast project, and we cannot guarantee that no power is exported. It's not even possible... If exporting power ist so critical, you need a piece of hardware in series with your power supply, which is able to react in realtime.

**Reply by rradkov** [reply]
> Sorry, but OpenDTU-OnBattery is a hobbyist/enthusiast project, and we cannot guarantee that no power is exported. It's not even possible... If exporting power ist so critical, you need a piece of hardware in series with your power supply, which is able to react in realtime.
>

Hey I know that this is hobbyist project and there is no guarantees but my main point was if those 3 switches: Intermediate meter POWERMETER_MIN_POINT ON_GRID_FEED_FAST_LIMIT_DECREASE are implemented as in the zero export script will probably solve the issue that potentially more people will encounter.

**Reply by schlimmchen** [reply]
Well... I am not going to research what these switches mean in that third party script... Maybe if you explain them well I can tell you if something similar is in OpenDTU-OnBattery or if I think it's worth a feature request.

**Reply by rradkov** [reply]
here is the information: you've asked::

[SELECT_INTERMEDIATE_METER]
# if you have an intermediate meter ("Zwischenzähler") to measure the outputpower of your inverter you can set it here. It is faster than the DTU current_power value
# --- define your intermediate meter - if you don´t have one set the following defines to false to use the value from your DTU---

# if your powermeter falls below POWERMETER_MIN_POINT: immediatelly decrease the limit
ON_GRID_FEED_FAST_LIMIT_DECREASE = false

# if your powermeter jumps under this point, the limit will be reduced instantly. it is like a "super high priority limit change".
POWERMETER_MIN_POINT = -600

**Reply by spcqike** [reply]
I don’t see any reason for these two switches. An intermediate meter might be faster, that’s true. But the Hoymiles is as slow as it is. You can’t transmit a new limit faster than it handles it anyway.

and with our actual DPL implementation I don’t see a reason for such a min point either as we have a target and an allowed hysteresis. Our cycle time is as fast as 1-2s. So if your load shuts down and you start exporting, the next DPL cycle will decrease your limit.

**Reply by Manos1966** [reply]
@rradkov I see you are still not gettting the point...

spcqike and schlimchen may correct me, your Problem is that the HOYMILES is too slow!  (not the OpenDTU-OnBattery).

Since you are really determined to find a solution, here is one option for you:

Radical Option: 
According to the number you provided (1,7kWh), if we suppose that the average electricity injection is at size I calculated in my other post, you have no batteries so it occurs only during daylight, you have on average SIX TIMES per day (otherwise SIX TIMES per 24hrs) a situation where HOYMILES produces more electricity than required.
SOLUTION: 
Install a SHELLY PM between HOYMILES and electricity (You already have it INSTALLED 😉 )
Install a SHELLY 3erm at Power Meter (You already have it INSTALLED 😉 )
Activate a condition on the Shelly 1PM: when the Shelly 3em is NEGATIVE by more than -xx, SHUTDOWN  HOYMILES for 1 Minute!
When the HOYMILES powers up, it increases production gradually thus takes time to reach the target level and by then, there is not energy injected into the network. 
This way you are safe 🥂 
  
Since you have everything needed already installed, it is very easy 😉

**Comment by Manos1966** [comment]
> Unfortunately here xxxx if you export energy to the grid, you need to follow some strict regulatory laws(only certified electrician can install PV systems) and sign a contract with electricity service provider. If PV system is not certified and you're exporting to the grid, I need to pay exported energy at the price of importing it. Also they can potentially give me a ticket for that.

Oh, I see!
The bad news: The HOYMILES are "slow" in adjusting their energy limit.
It is less of a question whether the OpenDTU-OnBattery is fast or not (of course you can improve it).
It is the issue of the HOYMILES receiving a command for a new Watt limit and how long it takes to adjust to it.

The "less bad" news: 

1.  We actually talk about small electricity volumes here
2. The amount of electricity that you inject to the grid is actually nuch less than you think:
    Your (and everybody's) graphs do NOT report with enough precision in order to give the correct picture.
    The electricity you pay in based on kWh = Kilowatts per hour = 1000Watts per 3600seconds
    Lets suppose your HOYMILES had -621,5 W (in your graph) at 52seconds and it took until 59second to correct it
    It took also 7 seconds to go from -621,5W up to 22W, a total of 643,5 W difference over 7 seconds
    Taking on a linear basis (for simplification - not true but, I am lazy today),  the rate is (643,5 / 7) 91,9W per second
    so:
    :52 second **-621,5** 
    :53 second -621,5 + 91,9 = **-529,6**
    :54 second -621,5 + 91,9 + 91,9 = **-437,7**
    :55 second -621,5 + 91,9 + 91,9 + 91,9 = **-345,8**
    :56 second -621,5 + 91,9 + 91,9 + 91,9 + 91,9 = **-253,9**
    :57 second -621,5 + 91,9 + 91,9 + 91,9 + 91,9 + 91,9 = **-162**
    :58 second -621,5 + 91,9 + 91,9 + 91,9 + 91,9 + 91,9 + 91,9 = **-70,1**
    :59 second +22
    
    So.... in total (621,5+529,6+437,7+345,8+253,9+162+70) = 2420,6 / 7 = you injected on average -345,8 per second (for seven seconds!
How many kWh it represents = (345,8 / 1000) x 7sec / 3600sec = **0,00067 kWh**
Medium size households reached a record high of EUR0.12 Kilowatt-hour in December of 2024 (I wish we had that price in Germany...)
So, EUR0,12 x 0,00067 kWh = my calculator can not even show this number! 😆 0,00008 EUR or 0,008cent

What I am trying to say, is: 
1. I do not believe your electricity provider will make an issue if you inject into the grid 0,00067 kWh every time your HOYMILES makes an adjustment.
2. If you have to pay for it, I do not believe you should worry yourself about 0,008cent
3. I would question if your electricity provider has the means to measure this UNLESS you have an electricity meter measuring the amount of electricity you inject back to the network (our German ones do that). EVEN IF THIS HAPPENS, I do not believe it will be an issue for the amount of electricity injected.

**Reply by rradkov** [reply]
> > Unfortunately here xxxx if you export energy to the grid, you need to follow some strict regulatory laws(only certified electrician can install PV systems) and sign a contract with electricity service provider. If PV system is not certified and you're exporting to the grid, I need to pay exported energy at the price of importing it. Also they can potentially give me a ticket for that.
> 
> Oh, I see! The bad news: The HOYMILES are "slow" in adjusting their energy limit. It is less of a question whether the OpenDTU-OnBattery is fast or not (of course you can improve it). It is the issue of the HOYMILES receiving a command for a new Watt limit and how long it takes to adjust to it.
> 
> The "less bad" news:
> 
> 1. We actually talk about small electricity volumes here
> 2. The amount of electricity that you inject to the grid is actually nuch less than you think:
>    Your (and everybody's) graphs do NOT report with enough precision in order to give the correct picture.
>    The electricity you pay in based on kWh = Kilowatts per hour = 1000Watts per 3600seconds
>    Lets suppose your HOYMILES had -621,5 W (in your graph) at 52seconds and it took until 59second to correct it
>    It took also 7 seconds to go from -621,5W up to 22W, a total of 643,5 W difference over 7 seconds
>    Taking on a linear basis (for simplification - not true but, I am lazy today),  the rate is (643,5 / 7) 91,9W per second
>    so:
>    :52 second **-621,5**
>    :53 second -621,5 + 91,9 = **-529,6**
>    :54 second -621,5 + 91,9 + 91,9 = **-437,7**
>    :55 second -621,5 + 91,9 + 91,9 + 91,9 = **-345,8**
>    :56 second -621,5 + 91,9 + 91,9 + 91,9 + 91,9 = **-253,9**
>    :57 second -621,5 + 91,9 + 91,9 + 91,9 + 91,9 + 91,9 = **-162**
>    :58 second -621,5 + 91,9 + 91,9 + 91,9 + 91,9 + 91,9 + 91,9 = **-70,1**
>    :59 second +22
>    So.... in total (621,5+529,6+437,7+345,8+253,9+162+70) = 2420,6 / 7 = you injected on average -345,8 per second (for seven seconds!
>    How many kWh it represents = (345,8 / 1000) x 7sec / 3600sec = **0,00067 kWh**
>    Medium size households reached a record high of EUR0.12 Kilowatt-hour in December of 2024 (I wish we had that price in Germany...)
>    So, EUR0,12 x 0,00067 kWh = my calculator can not even show this number! 😆 0,00008 EUR or 0,008cent
> 
> What I am trying to say, is:
> 
> 1. I do not believe your electricity provider will make an issue if you inject into the grid 0,00067 kWh every time your HOYMILES makes an adjustment.
> 2. If you have to pay for it, I do not believe you should worry yourself about 0,008cent
> 3. I would question if your electricity provider has the means to measure this UNLESS you have an electricity meter measuring the amount of electricity you inject back to the network (our German ones do that). EVEN IF THIS HAPPENS, I do not believe it will be an issue for the amount of electricity injected.

well you're focused on the single moment provided by me. If you look it at monthly base there are 1,7kW of energy which here is around 25 euro cents. As far I know the provided energy meter by the electricity company is counting the exported energy to the grid as well. Yeah if I need to pay it it doesn't bother me at all but the fine that I can potentially face is worrying me.
So as I mentioned previously if the developers manage to add those 3 settings - Intermediate meter, POWERMETER_MIN_POINT, ON_GRID_FEED_FAST_LIMIT_DECREASE - which are already defined in Zero Export script potentially this can resolve the issue with the exported energy.

**Reply by Manos1966** [reply]
I see. As I said, the response time is primarily a HOYMILES inverter issue and less an OpenDTUonBattery issue.

"the provided energy meter by the electricity company is counting the exported energy to the grid as well"
What is the Brand name and model of your electricity meter? Perhaps we can check it.

**Reply by rradkov** [reply]
> I see. As I said, the response time is primarily a HOYMILES inverter issue and less an OpenDTUonBattery issue.
> 
> "the provided energy meter by the electricity company is counting the exported energy to the grid as well" What is the Brand name and model of your electricity meter? Perhaps we can check it.

Hey Manos1966, thanks for the support !

**Reply by Manos1966** [reply]
Just a quick one: It is better to post pictures WITHOUT the Serial number of your electricity meter.
I do not really know why but, better please delete these pictures (one never knows)

Your model is a sanxing dlms S12U16
(and you have THREE ! 😨 )

And it is single phase meter so I gues they install three meters, one for each phase? 😨 🤦‍♂️

**Reply by rradkov** [reply]
well, the picture showed not only mine meter 👯

**Reply by Manos1966** [reply]
Well, unfortunately you are right, this meter does collect export data 😞 
Practically, if the meter shows something starting with 1.8. it shows your consumption. 
If it shows something with 2.8. it is what you export to the grid.

1.8.0 Energy import,kWh = what you are consuming
2.8.0 Energy export,kWh = what you give to the grid

and then there is more. 
1.8.1 Energy import T1,kWh
1.8.2 Energy import T2,kWh
2.8.1 Energy export T1,kWh
2.8.2 Energy export T2,kWh

I understand you challenge, there seems to be a long way to go in your country:

"In our country, the first photovoltaic system on a balcony was installed in a municipal housing in Dupnitsa. **The reason for this is primarily the lack of a legal framework and the still unregulated legal path.** 
**At the moment, it is easy to install such systems on private houses**, but this is not the case for multi-family residential buildings. 
There, it is necessary for the homeowner, who decides to put such a system on his balcony, to have the consent of 51% of the owners of the building. Permission from the energy company serving the building is also required."

"The first photovoltaic microsystem on a balcony is already a fact in Bulgaria. It was installed in a municipal housing, where the socially poor live, in the city of Dupnitsa. It is implemented under the Sunrise project – Households producing household energy from the sun with the support of Habitat Bulgaria and the municipality of Dupnitsa. The goal is for households living in social poverty, and subsequently all others, to reduce their electricity bills. Another advantage is environmental protection."

"The practice of placing solar panels on residential balconies is now firmly established in Germany. For this reason, the first installation was installed precisely by German experts from the German Solar Energy Society (DGS)."  

You may want to get a connect to this person who runs the SUNRISE Project:
The SUNRISE Project aims to spread and establish balcony power plants in Eastern and South-Eastern European countries.
Christoph Johan, photovoltaic project engineer from the German Solar Energy Society (DGS)
https://www.dgs-berlin.de/solarprojects/sunrise/?lang=en
He seems to work with Bulgaria Habitat for Humanity
https://www.habitat.org/emea/bg/city/bulgaria-habitat-humanity
so probably you could call them and ask them if they have found an easier way to install "balcony power plants" in Bulgaria.

https://fakti.bg/en/imoti/901060-they-installed-the-first-solar-panel-on-a-balcony-in-our-country-this-is-how-we-produce-our-own-electricity-video

---------Summarizing--------------------
- There are movements in your country to facilitate installation of balcony power plants (or plug-in solar devices)
- There is no legal framework (although this appears to be a major issue (as it is the case in Germany) for buldings with many home owners - Not for single homes
- There are some experts you can get in contact with to ask for more information

Considering everything I read, I am nevertheless surprised that the electricity companies are already installing such high-tech digital electricity meters which already have import/export measuring capability 👍

**Reply by Manos1966** [reply]
I do not believe you are going to have an issue with your electricity company because:
- You have a high-tech electricity meter = You do not "cheat" by making the electricity meter go "backwards" i.e. show less elecricity then what you actually consumed, which is =highly illegal (similar to the days people changed the odemeters of the cars to show less kilometers driven).
- You have a system installed to ensure "zero feed -in" 
- The amount of electricity you export is "accidental" and it is so small that it is irrelevant
- You pay (if I understand correctly) even for this - although I do not believe this to be the case with your electricity meter! This was the case for some electricity meters that would add Import and Export together to one number (so you would pay the sum).

But all this is theory... I do not live in Bulgaria so, better check. 🤞

**Reply by Manos1966** [reply]
> If exporting power ist so critical, you need a piece of hardware in series with your power supply, which is able to react in realtime.

I have a crazy idea which is more "opening Pandora's box" but I mention it for your future plans:
Do you have a 48V battery? 
You get a HUAWEI R4850 power supply (they used to cost second-hand 40 USD, now they are over 130EUR 🤮 )
The HUAWEI R4850 can deliver from 100W to 3000W charging power to the 48V battery.
It has response time (adjusting its power) of less than two seconds!
So, with a HUAWEI R4850 in your system, any excess electricity will be immediatelly absorbed by the HUAWEI ! 😎 
https://github.com/helgeerbe/OpenDTU-OnBattery/discussions/1239#discussioncomment-10682549

**Comment by Gururom** [comment]
Ich würde als erstes das minimum powerlimit im DPL höher setzen, so 40-50watt.
So wird die Einspeisung auch nochmal geringer

**Reply by rradkov** [reply]
I was thinking that this can be the problem, but it is not... The problem here is the slow reaction of reducing the output of the inverter. That's why I suggested 3 new parameters to be added to the DPL policy
